### ISSUE TITLE
Add support for Symfony >= 5.1 new Authenticator-based system

### ISSUE THREAD

**Issue by welcoMattic** [knpuniversity/oauth2-client-bundle#282]
Hi!

After watching again @weaverryan's keynote at SymfonyWorld (congrats, it was an amazing talk! :clap: ) I wondered if this bundle was or would be compatible with the new authenticator-based system documented [here](https://symfony.com/doc/current/security/experimental_authenticators.html) 

What would be the workload?
Is backwards compatibility possible? Or will this result in a major version of the bundle?

I'll be happy to contribute to the addition of this support. 

Thank you!

**Comment by weaverryan** [comment]
Hey @welcoMattic!

Thanks for the message! As you can see, as much as I love the new security system, I'm very short on time to even make this bundle compatible with it. So yes, I would love your help if you're able!

Related to security (specifically, the parts of security that *changed* with the new 5.2 mode), this bundle only has 1 class I believe - the authenticator: https://github.com/knpuniversity/oauth2-client-bundle/blob/master/src/Security/Authenticator/SocialAuthenticator.php

So, what we need to do is:

A) Keep the old class. We can remove it some time FAR into the future when Guard authentication no longer lives in Symfony.

B) Create a new authenticator class (what should we call it? `Social52Authenticator` seems silly... maybe `OAuth2Authenticator` just to have a different name) that basically does the same thing as the old system, but uses the new authenticator interface.

C) Update the docs! A big part of the README talks about how to use the `SocialAuthenticator`. We don't want to just delete the old code and replace it with the new code, as both are still valid. We may need to have two sections - one showing the old SocialAuthenticator and another showing the new authenticator class - but I'm not 100% sure what's best yet.

So yes,! If you're able, I would love help on this!

Cheers!

**Comment by Anthodev** [comment]
@weaverryan @jrushlow I did a first draft if you're interested, not sure it's entirely correct but it should give at least a support to make the job work, see #292

**Comment by soebbing** [comment]
Sorry to highjack this issue, but I thought since the PR was merged, I might ask some questions :)

I tried to implement the `FacebookAuthenticator` as described in the [Readme](https://github.com/knpuniversity/oauth2-client-bundle/tree/v2.8.0#using-the-new-symfony-authenticator), but the underlying interface seems off.

I get the error 
```
Symfony\Component\Security\Guard\Authenticator\GuardBridgeAuthenticator::__construct():
Argument #1 ($guard) must be of type Symfony\Component\Security\Guard\AuthenticatorInterface, App\Security\FacebookAuthenticator given
```
This seems to happen since the `FacebookAuthenticator` implements the `\Symfony\Component\Security\Http\Authenticator\AuthenticatorInterface`, not the `Symfony\Component\Security\Guard\AuthenticatorInterface`. Am I using the service wrong?

I also didn't quite understand the relation to the [Guard Authenticator](https://github.com/knpuniversity/oauth2-client-bundle/tree/v2.8.0#authenticating-with-guard) - are those for different underlying authentication mechanisms, pre and post Symfony 5.2? Or is something else going on?

Have a great day!

**Comment by weaverryan** [comment]
Hey @soebbing!

Hmm. It sounds like you're using the new system (the non-Guard system), but potentially are still listing your `FacebookAuthenticator` under `guard` in security.yaml. Is that accurate?

Basically, when using the new security system, Guard is completely replaced. However, we decided to support the `guard` key in the new system to help migration: you can have an old-style Guard authenticator (the type described here https://github.com/knpuniversity/oauth2-client-bundle/tree/v2.8.0#authenticating-with-guard ) and use it directly (with zero modifications) in the new security system, but sticking it under the `guard` key. Or, to say it differently, if you have a working Guard authenticator, which is (of course) listed under the `guard` key of your firewall, you can switch to the new security system (`enable_authenticator_manager: true`) without making ANY other changes: your guard authenticators WILL still work because of this `GuardBridgeAuthenticator` which takes the stuff under your `guard` key and "bridges" them to work with the new system.

The tl;dr is probably this. If you're using the new security system (`enable_authenticator_manager: true`) you have 2 options:

1) You CAN still write a guard authenticator and stick it under the `guard` key like before. But this is to help migrating, and it will probably be deprecated soon.

2) Write a new style authenticator (non-guard): https://github.com/knpuniversity/oauth2-client-bundle/tree/v2.8.0#using-the-new-symfony-authenticator - then stick it under the `custom_authenticators` key of your firewall.

It sounds to me (but tell me if I'm wrong) that you did (2), but then used it under the `guard` key.

Cheers!

**Comment by soebbing** [comment]
Hey @weaverryan!

First: Thank you very much for taking your time and writing such a thorough response, very much appreciated! :heart: 

Your guess was right on: I am using the new security system and had added the `FacebookAuthenticator` under the `guard` key in `security.yml`. For reasons that I am not really aware of, I used a custom guard service (which is the old way, as I learned now thanks to you) for the new login link authentication. Your explanations really helped me out there!

Moving the `FacebookAuthenticator` to the `custom_authenticators` key fixed these issues - authentication is working now! :tada: 

PS: There is a minor issue in the [OAuth2Authenticator example](https://github.com/knpuniversity/oauth2-client-bundle/#step-1-using-the-new-oauth2authenticator-class), preventing it from working directly (the class property `$entityManager` being defined in the constructor is called `$em` in the other methods), might I create a PR for that?

**Comment by weaverryan** [comment]
Yay! I'm glad we were able to sort it out.

And thanks for the pointer on the README - you were right - and I fixed it at sha: 92ecbc550ee76ab87b78b62eb6237f3733c2eead

Cheers!
